# Enhanced MLP Architecture + Hyperparameter Exploration Notebook

**What this notebook does**
- Loads a CSV file that is already *categorically encoded* (user must provide path).
- Tries a configurable range of MLP architectures and hyperparameters (hidden layers, solvers, batch sizes, learning rates, epochs, and — for PyTorch models — dropout).
- Runs experiments using both **scikit-learn MLP** (fast, good for many grid search experiments) and an optional **PyTorch MLP** (to test dropout explicitly and obtain training loss/accuracy curves).
- **Enhanced logging**: Comprehensive logging with TensorBoard, MLflow, progress tracking, error handling, and resource monitoring.
- Saves several graphs (PNG files) showing how different parameters affect training/test performance and training loss curves.
- Saves results to a CSV for later inspection.

**Notes**
- The notebook assumes the CSV includes features and a target column (default target column is `risk_level`). If your target has string labels, the notebook will attempt to map `low/medium/high` to `0/1/2`. Adjust the mapping if needed.
- The notebook uses `StandardScaler` to scale inputs (recommended for neural nets).
- The PyTorch implementation supports dropout and logs all metrics to TensorBoard and optionally MLflow.
- Enhanced with comprehensive logging, error handling, and resource monitoring.

You can run this notebook end-to-end; change the parameter grids near the top to expand or narrow the search.


In [14]:
# Enhanced imports with logging and monitoring
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from itertools import product
from datetime import datetime
import logging
import traceback
import psutil
import yaml
from tqdm import tqdm
import json

# ML imports
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import (accuracy_score, classification_report, confusion_matrix, 
                           precision_score, recall_score, f1_score, precision_recall_fscore_support)
import joblib  # For saving/loading models and data

# Set up enhanced logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('experiment.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

# Configuration
OUTDIR = 'experiment_outputs'
os.makedirs(OUTDIR, exist_ok=True)
os.makedirs(os.path.join(OUTDIR, 'models'), exist_ok=True)
os.makedirs(os.path.join(OUTDIR, 'tb_logs'), exist_ok=True)
os.makedirs(os.path.join(OUTDIR, 'sklearn_models'), exist_ok=True)
os.makedirs(os.path.join(OUTDIR, 'pytorch_models'), exist_ok=True)
os.makedirs(os.path.join(OUTDIR, 'metrics'), exist_ok=True)
os.makedirs(os.path.join(OUTDIR, 'confusion_matrices'), exist_ok=True)

# Experiment configuration
use_pytorch = True  # Set to False to skip PyTorch experiments
use_mlflow = True  # Set to True to enable MLflow logging
use_cross_validation = True  # Set to True to enable 5-fold cross-validation
n_folds = 5  # Number of cross-validation folds
data_path = '/mnt/data/sample.csv'  # Change this to your CSV path

logger.info(f"Starting ML experiments at {datetime.now()}")
logger.info(f"Output directory: {OUTDIR}")
logger.info(f"PyTorch enabled: {use_pytorch}")
logger.info(f"MLflow enabled: {use_mlflow}")
logger.info(f"Cross-validation enabled: {use_cross_validation} ({n_folds} folds)")

2025-09-06 00:12:30,876 - INFO - Starting ML experiments at 2025-09-06 00:12:30.876181
2025-09-06 00:12:30,877 - INFO - Output directory: experiment_outputs
2025-09-06 00:12:30,877 - INFO - PyTorch enabled: True
2025-09-06 00:12:30,878 - INFO - MLflow enabled: True
2025-09-06 00:12:30,878 - INFO - Cross-validation enabled: True (5 folds)
2025-09-06 00:12:30,877 - INFO - Output directory: experiment_outputs
2025-09-06 00:12:30,877 - INFO - PyTorch enabled: True
2025-09-06 00:12:30,878 - INFO - MLflow enabled: True
2025-09-06 00:12:30,878 - INFO - Cross-validation enabled: True (5 folds)


In [2]:
# Helper function for comprehensive metrics calculation
def calculate_comprehensive_metrics(y_true, y_pred, model_name, experiment_id, model_type='sklearn'):
    """Calculate and save comprehensive metrics for a model"""
    
    # Calculate all metrics
    accuracy = accuracy_score(y_true, y_pred)
    precision_macro = precision_score(y_true, y_pred, average='macro', zero_division=0)
    recall_macro = recall_score(y_true, y_pred, average='macro', zero_division=0)
    f1_macro = f1_score(y_true, y_pred, average='macro', zero_division=0)
    
    precision_weighted = precision_score(y_true, y_pred, average='weighted', zero_division=0)
    recall_weighted = recall_score(y_true, y_pred, average='weighted', zero_division=0)
    f1_weighted = f1_score(y_true, y_pred, average='weighted', zero_division=0)
    
    # Per-class metrics
    precision_per_class, recall_per_class, f1_per_class, support = precision_recall_fscore_support(
        y_true, y_pred, zero_division=0)
    
    # Confusion matrix
    cm = confusion_matrix(y_true, y_pred)
    
    # Create comprehensive metrics dictionary
    metrics = {
        'model_name': model_name,
        'experiment_id': experiment_id,
        'model_type': model_type,
        'accuracy': accuracy,
        'precision_macro': precision_macro,
        'recall_macro': recall_macro,
        'f1_macro': f1_macro,
        'precision_weighted': precision_weighted,
        'recall_weighted': recall_weighted,
        'f1_weighted': f1_weighted,
        'confusion_matrix': cm.tolist(),
        'per_class_metrics': {
            'precision': precision_per_class.tolist(),
            'recall': recall_per_class.tolist(),
            'f1_score': f1_per_class.tolist(),
            'support': support.tolist()
        }
    }
    
    # Save metrics to JSON
    metrics_file = os.path.join(OUTDIR, 'metrics', f'{model_type}_model_{experiment_id}_metrics.json')
    with open(metrics_file, 'w') as f:
        json.dump(metrics, f, indent=2)
    
    # Create and save confusion matrix plot
    plt.figure(figsize=(8, 6))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', 
                xticklabels=[f'Class {i}' for i in range(cm.shape[1])],
                yticklabels=[f'Class {i}' for i in range(cm.shape[0])])
    plt.title(f'Confusion Matrix - {model_name}')
    plt.xlabel('Predicted')
    plt.ylabel('Actual')
    
    cm_file = os.path.join(OUTDIR, 'confusion_matrices', f'{model_type}_model_{experiment_id}_confusion_matrix.png')
    plt.savefig(cm_file, dpi=300, bbox_inches='tight')
    plt.close()
    
    # Generate classification report
    class_report = classification_report(y_true, y_pred, output_dict=True, zero_division=0)
    report_file = os.path.join(OUTDIR, 'metrics', f'{model_type}_model_{experiment_id}_classification_report.json')
    with open(report_file, 'w') as f:
        json.dump(class_report, f, indent=2)
    
    logger.info(f"Comprehensive metrics saved for {model_name} (ID: {experiment_id})")
    
    return metrics

# Load and prepare data
try:
    logger.info(f"Loading data from {data_path}")
    df = pd.read_csv("training_data.csv")
    logger.info(f"Data loaded successfully. Shape: {df.shape}")
    logger.info(f"Columns: {list(df.columns)}")
    
    # Display basic info about the dataset
    print("Dataset Info:")
    print(f"Shape: {df.shape}")
    print(f"Columns: {list(df.columns)}")
    print(f"Missing values: {df.isnull().sum().sum()}")
    
except Exception as e:
    logger.error(f"Failed to load data: {str(e)}")
    logger.error(traceback.format_exc())
    raise

# Prepare features and target
target_col = 'risk_level'  # Change this if your target column has a different name

try:
    if target_col not in df.columns:
        logger.warning(f"Target column '{target_col}' not found. Available columns: {list(df.columns)}")
        target_col = input("Please enter the correct target column name: ")
    
    X = df.drop(columns=[target_col])
    y = df[target_col]
    
    # Handle categorical target if needed
    if y.dtype == 'object':
        logger.info(f"Converting categorical target. Unique values: {y.unique()}")
        if set(y.unique()).issubset({'low', 'medium', 'high'}):
            y = y.map({'low': 0, 'medium': 1, 'high': 2})
            logger.info("Mapped risk levels: low->0, medium->1, high->2")
        else:
            # Use label encoding for other categorical targets
            from sklearn.preprocessing import LabelEncoder
            le = LabelEncoder()
            y = le.fit_transform(y)
            logger.info(f"Label encoded target. Classes: {le.classes_}")
    
    logger.info(f"Target distribution: {np.bincount(y)}")
    
except Exception as e:
    logger.error(f"Error preparing target variable: {str(e)}")
    logger.error(traceback.format_exc())
    raise

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
logger.info(f"Train-test split completed. Train: {X_train.shape}, Test: {X_test.shape}")

# Scale features
scaler = StandardScaler()
X_train_s = scaler.fit_transform(X_train)
X_test_s = scaler.transform(X_test)
logger.info("Features scaled using StandardScaler")

2025-09-06 00:02:00,157 - INFO - Loading data from /mnt/data/sample.csv
2025-09-06 00:02:00,211 - INFO - Data loaded successfully. Shape: (25000, 69)
2025-09-06 00:02:00,211 - INFO - Columns: ['has_dental_data', 'has_dietary_data', 'sa_citizen', 'special_needs', 'caregiver_treatment', 'appliance', 'plaque', 'dry_mouth', 'enamel_defects', 'fluoride_water', 'fluoride_toothpaste', 'topical_fluoride', 'regular_checkups', 'sealed_pits', 'restorative_procedures', 'enamel_change', 'dentin_discoloration', 'white_spot_lesions', 'cavitated_lesions', 'multiple_restorations', 'missing_teeth', 'total_dmft_score', 'sweet_sugary_foods', 'sweet_sugary_foods_bedtime', 'sweet_sugary_foods_daily', 'sweet_sugary_foods_weekly', 'sweet_sugary_foods_timing', 'takeaways_processed_foods', 'takeaways_processed_foods_daily', 'takeaways_processed_foods_weekly', 'fresh_fruit', 'fresh_fruit_bedtime', 'fresh_fruit_daily', 'fresh_fruit_weekly', 'fresh_fruit_timing', 'cold_drinks_juices', 'cold_drinks_juices_bedtime',

Dataset Info:
Shape: (25000, 69)
Columns: ['has_dental_data', 'has_dietary_data', 'sa_citizen', 'special_needs', 'caregiver_treatment', 'appliance', 'plaque', 'dry_mouth', 'enamel_defects', 'fluoride_water', 'fluoride_toothpaste', 'topical_fluoride', 'regular_checkups', 'sealed_pits', 'restorative_procedures', 'enamel_change', 'dentin_discoloration', 'white_spot_lesions', 'cavitated_lesions', 'multiple_restorations', 'missing_teeth', 'total_dmft_score', 'sweet_sugary_foods', 'sweet_sugary_foods_bedtime', 'sweet_sugary_foods_daily', 'sweet_sugary_foods_weekly', 'sweet_sugary_foods_timing', 'takeaways_processed_foods', 'takeaways_processed_foods_daily', 'takeaways_processed_foods_weekly', 'fresh_fruit', 'fresh_fruit_bedtime', 'fresh_fruit_daily', 'fresh_fruit_weekly', 'fresh_fruit_timing', 'cold_drinks_juices', 'cold_drinks_juices_bedtime', 'cold_drinks_juices_daily', 'cold_drinks_juices_weekly', 'cold_drinks_juices_timing', 'processed_fruit', 'processed_fruit_bedtime', 'processed_fruit_

In [3]:
# Check GPU availability
try:
    import torch
    gpu_available = torch.cuda.is_available()
    print(f"CUDA available: {gpu_available}")
    if gpu_available:
        print(f"GPU device: {torch.cuda.get_device_name(0)}")
        print(f"GPU memory: {torch.cuda.get_device_properties(0).total_memory / 1024**3:.1f} GB")
    else:
        print("No GPU available - will use CPU for PyTorch experiments")
except ImportError:
    print("PyTorch not installed - only scikit-learn experiments will run (CPU only)")

print(f"\nNote: Scikit-learn experiments always use CPU only")
print(f"PyTorch experiments will use {'GPU' if gpu_available else 'CPU'}")

CUDA available: True
GPU device: NVIDIA GeForce RTX 4050 Laptop GPU
GPU memory: 6.0 GB

Note: Scikit-learn experiments always use CPU only
PyTorch experiments will use GPU


## PyTorch GPU Installation

Based on your NVIDIA GeForce RTX 4050 with CUDA 12.9, install PyTorch with GPU support using one of these commands:

### Option 1: Latest PyTorch with CUDA 12.1 (Recommended)
```bash
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121
```

### Option 2: Latest PyTorch with CUDA 11.8 (Alternative)
```bash
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
```

### Option 3: CPU-only version (if GPU setup fails)
```bash
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu
```

**Note:** The CUDA 12.1 version should work with your CUDA 12.9 driver (backwards compatible).

Run the cell below to check if GPU is detected after installation.

In [4]:
# Parameter grids

# sklearn_param_grid = {
#     'hidden_layer_sizes': [(25,), (32,), (64,)],
#     'solver': ['adam', 'lbfgs', 'adamw'],
#     'learning_rate_init': [0.01, 0.1],
#     'max_iter': [200, 500, 1000]
# }

sklearn_param_grid = {
    'hidden_layer_sizes': [(64,)],
    'solver': ['adamw'],
    'learning_rate_init': [0.01],
    'max_iter': [500]
}

# pytorch_param_grid = {
#     'hidden_sizes': [[25], [32], [64]],
#     'learning_rate': [0.001, 0.01, 0.1],
#     'dropout': [0.2, 0.5],
#     'batch_size': [64, 128],
#     'epochs': [100, 150]
# }

pytorch_param_grid = {
    'hidden_sizes': [[64]],
    'learning_rate': [0.001],
    'dropout': [0.2],
    'batch_size': [64, 128],
    'epochs': [100, 150]
}

# Save experiment configuration
experiment_config = {
    'sklearn_param_grid': sklearn_param_grid,
    'pytorch_param_grid': pytorch_param_grid,
    'data_path': data_path,
    'target_column': target_col,
    'timestamp': datetime.now().isoformat(),
    'train_size': X_train.shape[0],
    'test_size': X_test.shape[0],
    'n_features': X_train.shape[1],
    'n_classes': len(np.unique(y)),
    'use_pytorch': use_pytorch,
    'use_mlflow': use_mlflow
}

with open(os.path.join(OUTDIR, 'experiment_config.yaml'), 'w') as f:
    yaml.dump(experiment_config, f, default_flow_style=False)

logger.info(f"Experiment configuration saved to {os.path.join(OUTDIR, 'experiment_config.yaml')}")
print(f"Total sklearn combinations: {len(list(product(*sklearn_param_grid.values())))}")
if use_pytorch:
    print(f"Total PyTorch combinations: {len(list(product(*pytorch_param_grid.values())))}")

2025-09-06 00:02:02,606 - INFO - Experiment configuration saved to experiment_outputs\experiment_config.yaml


Total sklearn combinations: 1
Total PyTorch combinations: 4


In [5]:
# Feature Selection - Use only the selected features from feature selection analysis
selected_features = [
    'has_dental_data',
    'special_needs',
    'caregiver_treatment',
    'plaque',
    'dry_mouth',
    'enamel_defects',
    'fluoride_water',
    'fluoride_toothpaste',
    'topical_fluoride',
    'regular_checkups',
    'sealed_pits',
    'enamel_change',
    'dentin_discoloration',
    'white_spot_lesions',
    'cavitated_lesions',
    'multiple_restorations',
    'missing_teeth',
    'total_dmft_score',
    'sweet_sugary_foods',
    'sweet_sugary_foods_bedtime',
    'takeaways_processed_foods',
    'fresh_fruit',
    'cold_drinks_juices',
    'processed_fruit',
    'processed_fruit_bedtime',
    'spreads',
    'spreads_bedtime',
    'added_sugars',
    'added_sugars_bedtime',
    'salty_snacks',
    'dairy_products',
    'vegetables',
    'has_dietary_data',
    'cold_drinks_juices_bedtime',
    'water'
]

logger.info(f"Selected {len(selected_features)} features for training")
logger.info(f"Selected features: {selected_features}")

# Check which selected features are available in the dataset
available_features = []
missing_features = []

for feature in selected_features:
    if feature in X.columns:
        available_features.append(feature)
    else:
        missing_features.append(feature)

logger.info(f"Available features: {len(available_features)}/{len(selected_features)}")
if missing_features:
    logger.warning(f"Missing features: {missing_features}")

# Filter X to only include selected features that are available
X_selected = X[available_features].copy()
logger.info(f"Feature matrix shape after selection: {X_selected.shape}")

# Update X to use only selected features
X = X_selected

print(f"Features selected: {len(available_features)}")
print(f"Final feature set: {list(X.columns)}")
if missing_features:
    print(f"Warning: {len(missing_features)} features not found in dataset: {missing_features}")

2025-09-06 00:02:02,615 - INFO - Selected 35 features for training
2025-09-06 00:02:02,616 - INFO - Selected features: ['has_dental_data', 'special_needs', 'caregiver_treatment', 'plaque', 'dry_mouth', 'enamel_defects', 'fluoride_water', 'fluoride_toothpaste', 'topical_fluoride', 'regular_checkups', 'sealed_pits', 'enamel_change', 'dentin_discoloration', 'white_spot_lesions', 'cavitated_lesions', 'multiple_restorations', 'missing_teeth', 'total_dmft_score', 'sweet_sugary_foods', 'sweet_sugary_foods_bedtime', 'takeaways_processed_foods', 'fresh_fruit', 'cold_drinks_juices', 'processed_fruit', 'processed_fruit_bedtime', 'spreads', 'spreads_bedtime', 'added_sugars', 'added_sugars_bedtime', 'salty_snacks', 'dairy_products', 'vegetables', 'has_dietary_data', 'cold_drinks_juices_bedtime', 'water']
2025-09-06 00:02:02,616 - INFO - Available features: 35/35
2025-09-06 00:02:02,621 - INFO - Feature matrix shape after selection: (25000, 35)
2025-09-06 00:02:02,616 - INFO - Selected features: ['h

Features selected: 35
Final feature set: ['has_dental_data', 'special_needs', 'caregiver_treatment', 'plaque', 'dry_mouth', 'enamel_defects', 'fluoride_water', 'fluoride_toothpaste', 'topical_fluoride', 'regular_checkups', 'sealed_pits', 'enamel_change', 'dentin_discoloration', 'white_spot_lesions', 'cavitated_lesions', 'multiple_restorations', 'missing_teeth', 'total_dmft_score', 'sweet_sugary_foods', 'sweet_sugary_foods_bedtime', 'takeaways_processed_foods', 'fresh_fruit', 'cold_drinks_juices', 'processed_fruit', 'processed_fruit_bedtime', 'spreads', 'spreads_bedtime', 'added_sugars', 'added_sugars_bedtime', 'salty_snacks', 'dairy_products', 'vegetables', 'has_dietary_data', 'cold_drinks_juices_bedtime', 'water']


In [6]:
# Scikit-learn MLP experiments with Cross-Validation and Comprehensive Metrics
sklearn_results = []
combos = list(product(*sklearn_param_grid.values()))
keys = list(sklearn_param_grid.keys())

logger.info(f"Starting {len(combos)} scikit-learn experiments")
if use_cross_validation:
    logger.info(f"Using {n_folds}-fold cross-validation (this will increase training time by ~{n_folds}x)")

# Combine train and test data for cross-validation
X_full = np.vstack([X_train_s, X_test_s])
y_full = np.concatenate([y_train, y_test])

# Set up cross-validation
if use_cross_validation:
    cv = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state=42)
    cv_splits = list(cv.split(X_full, y_full))

best_sklearn_model = None
best_sklearn_score = 0.0

for i, combo in enumerate(tqdm(combos, desc="Scikit-learn experiments")):
    try:
        params = dict(zip(keys, combo))
        logger.debug(f"Testing sklearn params: {params}")
        
        # Log system resources
        memory_usage = psutil.virtual_memory().percent
        cpu_usage = psutil.cpu_percent(interval=1)
        
        start_time = datetime.now()
        
        if use_cross_validation:
            # Cross-validation approach
            clf = MLPClassifier(**params, random_state=42)
            cv_scores = cross_val_score(clf, X_full, y_full, cv=cv, scoring='accuracy')
            
            # Train on full dataset to get other metrics
            clf.fit(X_full, y_full)
            
            # Calculate metrics on original train/test split for comparison
            train_pred = clf.predict(X_train_s)
            test_pred = clf.predict(X_test_s)
            train_acc = accuracy_score(y_train, train_pred)
            test_acc = accuracy_score(y_test, test_pred)
            
            cv_mean = cv_scores.mean()
            cv_std = cv_scores.std()
            
        else:
            # Original single train/test split approach
            clf = MLPClassifier(**params, random_state=42)
            clf.fit(X_train_s, y_train)
            
            train_pred = clf.predict(X_train_s)
            test_pred = clf.predict(X_test_s)
            train_acc = accuracy_score(y_train, train_pred)
            test_acc = accuracy_score(y_test, test_pred)
            
            cv_mean = None
            cv_std = None
        
        end_time = datetime.now()
        duration = (end_time - start_time).total_seconds()
        
        # Calculate comprehensive metrics for test set
        model_name = f"sklearn_mlp_{i+1}"
        test_metrics = calculate_comprehensive_metrics(
            y_test, test_pred, model_name, i+1, 'sklearn'
        )
        
        # Calculate comprehensive metrics for training set
        train_metrics = calculate_comprehensive_metrics(
            y_train, train_pred, f"{model_name}_train", f"{i+1}_train", 'sklearn'
        )
        
        # Save model
        import joblib
        model_path = os.path.join(OUTDIR, 'sklearn_models', f'sklearn_model_{i+1}.joblib')
        joblib.dump(clf, model_path)
        
        # Track best model
        score_for_comparison = cv_mean if use_cross_validation else test_acc
        if score_for_comparison > best_sklearn_score:
            best_sklearn_score = score_for_comparison
            best_sklearn_model = {
                'model': clf,
                'experiment_id': i+1,
                'params': params,
                'test_metrics': test_metrics,
                'train_metrics': train_metrics,
                'model_path': model_path
            }
        
        result = {
            'experiment_id': i + 1,
            'params': params,
            'train_acc': train_acc,
            'test_acc': test_acc,
            'train_precision_macro': train_metrics['precision_macro'],
            'train_recall_macro': train_metrics['recall_macro'],
            'train_f1_macro': train_metrics['f1_macro'],
            'test_precision_macro': test_metrics['precision_macro'],
            'test_recall_macro': test_metrics['recall_macro'],
            'test_f1_macro': test_metrics['f1_macro'],
            'test_precision_weighted': test_metrics['precision_weighted'],
            'test_recall_weighted': test_metrics['recall_weighted'],
            'test_f1_weighted': test_metrics['f1_weighted'],
            'cv_mean_acc': cv_mean,
            'cv_std_acc': cv_std,
            'n_iter': clf.n_iter_,
            'loss_curve': clf.loss_curve_ if hasattr(clf, 'loss_curve_') else None,
            'duration_seconds': duration,
            'memory_usage_percent': memory_usage,
            'cpu_usage_percent': cpu_usage,
            'timestamp': start_time.isoformat(),
            'model_path': model_path
        }
        
        sklearn_results.append(result)
        
        if use_cross_validation:
            logger.info(f"Sklearn experiment {i+1}/{len(combos)} completed. CV mean: {cv_mean:.4f}±{cv_std:.4f}, Test F1: {test_metrics['f1_macro']:.4f}")
        else:
            logger.info(f"Sklearn experiment {i+1}/{len(combos)} completed. Train acc: {train_acc:.4f}, Test F1: {test_metrics['f1_macro']:.4f}")
        
    except Exception as e:
        logger.error(f"Sklearn experiment {i+1} failed with params {params}: {str(e)}")
        logger.error(traceback.format_exc())
        continue

logger.info(f"Completed {len(sklearn_results)} successful scikit-learn experiments")

# Save best sklearn model with special designation
if best_sklearn_model:
    best_model_dir = os.path.join(OUTDIR, 'best_models')
    os.makedirs(best_model_dir, exist_ok=True)
    
    # Save best model
    best_model_path = os.path.join(best_model_dir, 'best_sklearn_model.joblib')
    joblib.dump(best_sklearn_model['model'], best_model_path)
    
    # Save best model metrics with special naming
    best_test_metrics = calculate_comprehensive_metrics(
        y_test, best_sklearn_model['model'].predict(X_test_s), 
        'BEST_sklearn_model', 'BEST', 'sklearn_best'
    )
    
    # Save best model info
    best_info = {
        'experiment_id': best_sklearn_model['experiment_id'],
        'params': best_sklearn_model['params'],
        'metrics': best_test_metrics,
        'model_path': best_model_path,
        'timestamp': datetime.now().isoformat()
    }
    
    with open(os.path.join(best_model_dir, 'best_sklearn_model_info.json'), 'w') as f:
        json.dump(best_info, f, indent=2)
    
    logger.info(f"Best sklearn model saved (Experiment {best_sklearn_model['experiment_id']}) with score: {best_sklearn_score:.4f}")

# Save sklearn results
if sklearn_results:
    sklearn_df = pd.DataFrame([
        {
            'experiment_id': r['experiment_id'],
            'hidden_layer_sizes': str(r['params']['hidden_layer_sizes']),
            'solver': r['params']['solver'],
            'learning_rate_init': r['params']['learning_rate_init'],
            'max_iter': r['params']['max_iter'],
            'train_acc': r['train_acc'],
            'test_acc': r['test_acc'],
            'train_f1_macro': r['train_f1_macro'],
            'test_f1_macro': r['test_f1_macro'],
            'test_precision_macro': r['test_precision_macro'],
            'test_recall_macro': r['test_recall_macro'],
            'test_f1_weighted': r['test_f1_weighted'],
            'cv_mean_acc': r['cv_mean_acc'],
            'cv_std_acc': r['cv_std_acc'],
            'n_iter': r['n_iter'],
            'duration_seconds': r['duration_seconds'],
            'memory_usage_percent': r['memory_usage_percent'],
            'cpu_usage_percent': r['cpu_usage_percent'],
            'timestamp': r['timestamp'],
            'model_path': r['model_path']
        } for r in sklearn_results
    ])
    sklearn_df.to_csv(os.path.join(OUTDIR, 'sklearn_results_summary.csv'), index=False)
    logger.info("Scikit-learn results saved to sklearn_results_summary.csv")
    
    # Display best results
    if use_cross_validation:
        best_sklearn = sklearn_df.loc[sklearn_df['cv_mean_acc'].idxmax()]
        print(f"\nBest scikit-learn result (by CV score):")
        print(f"CV accuracy: {best_sklearn['cv_mean_acc']:.4f}±{best_sklearn['cv_std_acc']:.4f}")
        print(f"Test accuracy: {best_sklearn['test_acc']:.4f}")
        print(f"Test F1-macro: {best_sklearn['test_f1_macro']:.4f}")
        print(f"Test Precision-macro: {best_sklearn['test_precision_macro']:.4f}")
        print(f"Test Recall-macro: {best_sklearn['test_recall_macro']:.4f}")
        print(f"Parameters: hidden_sizes={best_sklearn['hidden_layer_sizes']}, solver={best_sklearn['solver']}, lr={best_sklearn['learning_rate_init']}")
    else:
        best_sklearn = sklearn_df.loc[sklearn_df['test_acc'].idxmax()]
        print(f"\nBest scikit-learn result:")
        print(f"Test accuracy: {best_sklearn['test_acc']:.4f}")
        print(f"Test F1-macro: {best_sklearn['test_f1_macro']:.4f}")
        print(f"Test Precision-macro: {best_sklearn['test_precision_macro']:.4f}")
        print(f"Test Recall-macro: {best_sklearn['test_recall_macro']:.4f}")
        print(f"Parameters: hidden_sizes={best_sklearn['hidden_layer_sizes']}, solver={best_sklearn['solver']}, lr={best_sklearn['learning_rate_init']}")

2025-09-06 00:02:02,641 - INFO - Starting 1 scikit-learn experiments
2025-09-06 00:02:02,642 - INFO - Using 5-fold cross-validation (this will increase training time by ~5x)
Scikit-learn experiments:   0%|          | 0/1 [00:00<?, ?it/s]2025-09-06 00:02:03,696 - ERROR - Sklearn experiment 1 failed with params {'hidden_layer_sizes': (64,), 'solver': 'adamw', 'learning_rate_init': 0.01, 'max_iter': 500}: 
All the 5 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
5 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\vhuta\Desktop\OralSmart - Data Analysis\venv\Lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File

In [7]:
# Add this cell to check PyTorch installation
try:
    import torch
    print(f"✅ PyTorch installed: {torch.__version__}")
    print(f"✅ CUDA available: {torch.cuda.is_available()}")
    if torch.cuda.is_available():
        print(f"✅ GPU device: {torch.cuda.get_device_name(0)}")
    else:
        print("⚠️  CUDA not available - will use CPU")
except ImportError as e:
    print(f"❌ PyTorch not installed: {e}")
    print("Install with: pip install torch torchvision torchaudio")
except Exception as e:
    print(f"❌ PyTorch error: {e}")

✅ PyTorch installed: 2.8.0+cu128
✅ CUDA available: True
✅ GPU device: NVIDIA GeForce RTX 4050 Laptop GPU


In [8]:
# PyTorch experiments setup
if use_pytorch:
    try:
        import torch
        import torch.nn as nn
        import torch.optim as optim
        from torch.utils.data import TensorDataset, DataLoader
        from torch.utils.tensorboard import SummaryWriter
        
        # MLflow setup (optional)
        if use_mlflow:
            import mlflow
            import mlflow.pytorch
            mlflow.set_experiment("MLP_Hyperparameter_Search")
            logger.info("MLflow experiment tracking enabled")
        
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        logger.info(f'PyTorch device: {device}')
        
        if torch.cuda.is_available():
            logger.info(f'CUDA device: {torch.cuda.get_device_name(0)}')
            logger.info(f'CUDA memory: {torch.cuda.get_device_properties(0).total_memory / 1024**3:.1f} GB')
        
        def build_pytorch_model(input_size, hidden_sizes, dropout=0.0, n_classes=None):
            layers = []
            in_size = input_size
            for h in hidden_sizes:
                layers.append(nn.Linear(in_size, h))
                layers.append(nn.ReLU())
                if dropout and dropout > 0.0:
                    layers.append(nn.Dropout(dropout))
                in_size = h
            layers.append(nn.Linear(in_size, n_classes))
            return nn.Sequential(*layers)
        
        # Prepare PyTorch tensors
        X_train_t = torch.FloatTensor(X_train_s)
        y_train_t = torch.LongTensor(y_train.values if hasattr(y_train, 'values') else y_train)
        X_test_t = torch.FloatTensor(X_test_s)
        y_test_t = torch.LongTensor(y_test.values if hasattr(y_test, 'values') else y_test)
        
        logger.info("PyTorch setup completed successfully")
        
    except ImportError as e:
        logger.error(f"PyTorch import failed: {str(e)}")
        use_pytorch = False
    except Exception as e:
        logger.error(f"PyTorch setup failed: {str(e)}")
        logger.error(traceback.format_exc())
        use_pytorch = False
else:
    logger.info("PyTorch experiments disabled")

2025-09-06 00:02:05,088 - INFO - MLflow experiment tracking enabled
2025-09-06 00:02:05,089 - INFO - PyTorch device: cuda
2025-09-06 00:02:05,089 - INFO - CUDA device: NVIDIA GeForce RTX 4050 Laptop GPU
2025-09-06 00:02:05,090 - INFO - CUDA memory: 6.0 GB
2025-09-06 00:02:05,093 - INFO - PyTorch setup completed successfully
2025-09-06 00:02:05,089 - INFO - PyTorch device: cuda
2025-09-06 00:02:05,089 - INFO - CUDA device: NVIDIA GeForce RTX 4050 Laptop GPU
2025-09-06 00:02:05,090 - INFO - CUDA memory: 6.0 GB
2025-09-06 00:02:05,093 - INFO - PyTorch setup completed successfully


In [9]:
# Early Stopping Class for PyTorch Training
class EarlyStopping:
    """Early stops the training if validation loss doesn't improve after a given patience."""
    def __init__(self, patience=7, min_delta=0.0001, restore_best_weights=True, verbose=True):
        """
        Args:
            patience (int): How long to wait after last time validation loss improved.
                          Default: 7
            min_delta (float): Minimum change in the monitored quantity to qualify as an improvement.
                              Default: 0.0001
            restore_best_weights (bool): Whether to restore model weights from the epoch with the best value.
                                       Default: True
            verbose (bool): If True, prints a message for each validation loss improvement. 
                          Default: True
        """
        self.patience = patience
        self.min_delta = min_delta
        self.restore_best_weights = restore_best_weights
        self.verbose = verbose
        self.counter = 0
        self.best_loss = None
        self.early_stop = False
        self.val_loss_min = float('inf')
        self.best_model_state = None

    def __call__(self, val_loss, model):
        if self.best_loss is None:
            self.best_loss = val_loss
            self.save_checkpoint(val_loss, model)
        elif val_loss < self.best_loss - self.min_delta:
            self.best_loss = val_loss
            self.save_checkpoint(val_loss, model)
            self.counter = 0
        else:
            self.counter += 1
            if self.verbose:
                logger.debug(f'EarlyStopping counter: {self.counter} out of {self.patience}')
            if self.counter >= self.patience:
                self.early_stop = True

    def save_checkpoint(self, val_loss, model):
        '''Saves model when validation loss decrease.'''
        if self.verbose:
            logger.debug(f'Validation loss decreased ({self.val_loss_min:.6f} --> {val_loss:.6f}).  Saving model checkpoint...')
        self.val_loss_min = val_loss
        if self.restore_best_weights:
            self.best_model_state = model.state_dict().copy()

    def restore_best_model(self, model):
        """Restore the best model weights if available"""
        if self.restore_best_weights and self.best_model_state is not None:
            model.load_state_dict(self.best_model_state)
            if self.verbose:
                logger.info(f'Restored model to best validation loss: {self.val_loss_min:.6f}')

# Early stopping configuration
EARLY_STOPPING_PATIENCE = 10  # Stop if no improvement for 10 epochs
EARLY_STOPPING_MIN_DELTA = 0.0001  # Minimum improvement threshold

logger.info(f"Early stopping configured with patience={EARLY_STOPPING_PATIENCE}, min_delta={EARLY_STOPPING_MIN_DELTA}")

2025-09-06 00:02:05,104 - INFO - Early stopping configured with patience=10, min_delta=0.0001


## Early Stopping Implementation ✅

**Early stopping has been implemented** with the following features:

### **Configuration:**
- **Patience**: 10 epochs (stops training if no improvement for 10 consecutive epochs)
- **Minimum Delta**: 0.0001 (minimum improvement threshold to be considered significant)
- **Restore Best Weights**: Automatically restores model to the best validation loss checkpoint

### **Benefits:**
1. **Prevents Overfitting**: Stops training when validation loss stops improving
2. **Saves Training Time**: Automatically stops when further training won't help
3. **Optimal Performance**: Returns the model with the best validation performance
4. **Literature Recommended**: Early stopping is a standard best practice

### **Training Process:**
- **Cross-validation**: Uses reduced patience (5 epochs) for faster fold training
- **Final model**: Uses full patience (10 epochs) for optimal performance
- **Validation frequency**: Every epoch (required for effective early stopping)

### **Result Tracking:**
The results now include early stopping metrics:
- `early_stopped`: Whether early stopping was triggered
- `epochs_trained`: Actual epochs trained (may be less than specified)
- `best_val_loss`: Best validation loss achieved
- `early_stop_epoch`: Epoch where early stopping occurred

This implementation follows **literature best practices** and is especially beneficial for synthetic data where clear patterns allow for early convergence.

In [10]:
# Enhanced PyTorch experiments with Cross-Validation and Comprehensive Metrics
if use_pytorch:
    pytorch_results = []
    combos = list(product(*pytorch_param_grid.values()))
    keys = list(pytorch_param_grid.keys())
    
    logger.info(f"Starting {len(combos)} PyTorch experiments")
    if use_cross_validation:
        logger.info(f"Using {n_folds}-fold cross-validation (this will increase training time by ~{n_folds}x)")
    
    run_counter = 0
    best_test_acc = 0.0
    best_pytorch_model = None
    
    # Prepare full dataset for cross-validation
    X_full_t = torch.FloatTensor(X_full)
    y_full_t = torch.LongTensor(y_full)
    
    for combo in tqdm(combos, desc="PyTorch experiments"):
        run_counter += 1
        
        try:
            params = dict(zip(keys, combo))
            logger.info(f"\nPyTorch experiment {run_counter}/{len(combos)}: {params}")
            
            # MLflow run start
            if use_mlflow:
                mlflow.start_run()
                mlflow.log_params(params)
            
            # Cross-validation scores storage
            cv_train_accs = []
            cv_val_accs = []
            cv_final_losses = []
            
            if use_cross_validation:
                # Cross-validation loop
                for fold, (train_idx, val_idx) in enumerate(cv_splits):
                    logger.debug(f"Training fold {fold + 1}/{n_folds}")
                    
                    # Split data for this fold
                    X_fold_train = X_full_t[train_idx]
                    y_fold_train = y_full_t[train_idx]
                    X_fold_val = X_full_t[val_idx]
                    y_fold_val = y_full_t[val_idx]
                    
                    # Model setup for this fold
                    input_size = X_full.shape[1]
                    n_classes = len(np.unique(y_full))
                    model = build_pytorch_model(
                        input_size, 
                        params['hidden_sizes'], 
                        dropout=params['dropout'], 
                        n_classes=n_classes
                    ).to(device)
                    
                    criterion = nn.CrossEntropyLoss()
                    optimizer = optim.Adam(model.parameters(), lr=params['learning_rate'])
                    
                    # Data loader for this fold
                    fold_train_ds = TensorDataset(X_fold_train, y_fold_train)
                    fold_train_loader = DataLoader(fold_train_ds, batch_size=params['batch_size'], shuffle=True)
                    
                    # Training loop for this fold with early stopping
                    fold_early_stopping = EarlyStopping(
                        patience=EARLY_STOPPING_PATIENCE//2,  # Reduced patience for CV folds
                        min_delta=EARLY_STOPPING_MIN_DELTA,
                        verbose=False  # Less verbose for CV folds
                    )
                    
                    for epoch in range(params['epochs']):
                        model.train()
                        epoch_losses = []
                        
                        for batch_idx, (xb, yb) in enumerate(fold_train_loader):
                            xb = xb.to(device)
                            yb = yb.to(device)
                            
                            optimizer.zero_grad()
                            out = model(xb)
                            loss = criterion(out, yb)
                            loss.backward()
                            optimizer.step()
                            
                            epoch_losses.append(loss.item())
                        
                        # Validation on fold validation set every epoch for early stopping
                        model.eval()
                        with torch.no_grad():
                            fold_val_out = model(X_fold_val.to(device))
                            fold_val_loss = criterion(fold_val_out, y_fold_val.to(device))
                        
                        # Early stopping check
                        fold_early_stopping(fold_val_loss.item(), model)
                        
                        if fold_early_stopping.early_stop:
                            logger.debug(f"Fold {fold+1} early stopping at epoch {epoch+1}")
                            break
                    
                    # Restore best model for this fold
                    if fold_early_stopping.restore_best_weights:
                        fold_early_stopping.restore_best_model(model)
                    
                    # Evaluate this fold
                    model.eval()
                    with torch.no_grad():
                        # Training accuracy for this fold
                        fold_train_out = model(X_fold_train.to(device))
                        fold_train_pred = fold_train_out.argmax(dim=1).cpu().numpy()
                        fold_train_acc = accuracy_score(y_fold_train.cpu().numpy(), fold_train_pred)
                        
                        # Validation accuracy for this fold
                        fold_val_out = model(X_fold_val.to(device))
                        fold_val_pred = fold_val_out.argmax(dim=1).cpu().numpy()
                        fold_val_acc = accuracy_score(y_fold_val.cpu().numpy(), fold_val_pred)
                        fold_val_loss = criterion(fold_val_out, y_fold_val.to(device)).item()
                    
                    cv_train_accs.append(fold_train_acc)
                    cv_val_accs.append(fold_val_acc)
                    cv_final_losses.append(fold_val_loss)
                    
                    logger.debug(f"Fold {fold + 1}: train_acc={fold_train_acc:.4f}, val_acc={fold_val_acc:.4f}")
                
                # Calculate cross-validation statistics
                cv_train_mean = np.mean(cv_train_accs)
                cv_train_std = np.std(cv_train_accs)
                cv_val_mean = np.mean(cv_val_accs)
                cv_val_std = np.std(cv_val_accs)
                cv_loss_mean = np.mean(cv_final_losses)
            
            else:
                # No cross-validation
                cv_train_mean = cv_train_std = cv_val_mean = cv_val_std = cv_loss_mean = None
            
            # Train final model on original train/test split
            model = build_pytorch_model(
                X_train_s.shape[1], 
                params['hidden_sizes'], 
                dropout=params['dropout'], 
                n_classes=len(np.unique(y_train))
            ).to(device)
            
            criterion = nn.CrossEntropyLoss()
            optimizer = optim.Adam(model.parameters(), lr=params['learning_rate'])
            
            train_ds = TensorDataset(X_train_t, y_train_t)
            train_loader = DataLoader(train_ds, batch_size=params['batch_size'], shuffle=True)
            
            # Setup logging for final model
            log_dir = os.path.join(OUTDIR, f'tb_logs/run_{run_counter}')
            writer = SummaryWriter(log_dir=log_dir)
            
            train_losses = []
            val_losses = []
            val_accuracies = []
            start_time = datetime.now()
            
            # Training loop for final model with early stopping
            early_stopping = EarlyStopping(
                patience=EARLY_STOPPING_PATIENCE,
                min_delta=EARLY_STOPPING_MIN_DELTA,
                verbose=True
            )
            
            val_epochs = []  # Track which epochs we validated
            
            for epoch in range(params['epochs']):
                model.train()
                epoch_losses = []
                
                for batch_idx, (xb, yb) in enumerate(train_loader):
                    xb = xb.to(device)
                    yb = yb.to(device)
                    
                    optimizer.zero_grad()
                    out = model(xb)
                    loss = criterion(out, yb)
                    loss.backward()
                    optimizer.step()
                    
                    epoch_losses.append(loss.item())
                
                avg_train_loss = np.mean(epoch_losses)
                train_losses.append(avg_train_loss)
                
                # Validation every epoch for early stopping
                model.eval()
                with torch.no_grad():
                    val_out = model(X_test_t.to(device))
                    val_loss = criterion(val_out, y_test_t.to(device))
                    val_pred = val_out.argmax(dim=1).cpu().numpy()
                    val_acc = accuracy_score(y_test, val_pred)
                
                val_losses.append(val_loss.item())
                val_accuracies.append(val_acc)
                val_epochs.append(epoch)
                
                # Early stopping check
                early_stopping(val_loss.item(), model)
                
                # Logging every 5 epochs or at early stop
                if epoch % 5 == 0 or epoch == params['epochs'] - 1 or early_stopping.early_stop:
                    # TensorBoard logging
                    writer.add_scalar('Loss/train', avg_train_loss, epoch)
                    writer.add_scalar('Loss/validation', val_loss.item(), epoch)
                    writer.add_scalar('Accuracy/validation', val_acc, epoch)
                    
                    # MLflow logging
                    if use_mlflow:
                        mlflow.log_metric("train_loss", avg_train_loss, step=epoch)
                        mlflow.log_metric("val_loss", val_loss.item(), step=epoch)
                        mlflow.log_metric("val_accuracy", val_acc, step=epoch)
                    
                    # System resource monitoring
                    memory_usage = psutil.virtual_memory().percent
                    if torch.cuda.is_available():
                        gpu_memory = torch.cuda.memory_allocated() / 1024**3
                        writer.add_scalar('System/gpu_memory_gb', gpu_memory, epoch)
                    writer.add_scalar('System/memory_usage_percent', memory_usage, epoch)
                
                # Progress reporting
                if (epoch + 1) % 20 == 0 or epoch == 0 or early_stopping.early_stop:
                    logger.debug(f"Epoch {epoch+1}/{params['epochs']}: train_loss={avg_train_loss:.4f}, val_loss={val_loss.item():.4f}")
                
                # Early stopping break
                if early_stopping.early_stop:
                    logger.info(f"Early stopping triggered at epoch {epoch+1}/{params['epochs']} (patience={EARLY_STOPPING_PATIENCE})")
                    logger.info(f"Best validation loss: {early_stopping.val_loss_min:.6f}")
                    break
            
            # Restore best model if early stopping was used
            if early_stopping.restore_best_weights:
                early_stopping.restore_best_model(model)
                logger.info("Model weights restored to best validation loss checkpoint")
            
            # Final evaluation
            model.eval()
            with torch.no_grad():
                out_train = model(X_train_t.to(device))
                pred_train = out_train.argmax(dim=1).cpu().numpy()
                out_test = model(X_test_t.to(device))
                pred_test = out_test.argmax(dim=1).cpu().numpy()
            
            train_acc = accuracy_score(y_train, pred_train)
            test_acc = accuracy_score(y_test, pred_test)
            
            end_time = datetime.now()
            duration = (end_time - start_time).total_seconds()
            
            # Calculate comprehensive metrics
            model_name = f"pytorch_mlp_{run_counter}"
            test_metrics = calculate_comprehensive_metrics(
                y_test, pred_test, model_name, run_counter, 'pytorch'
            )
            
            train_metrics = calculate_comprehensive_metrics(
                y_train, pred_train, f"{model_name}_train", f"{run_counter}_train", 'pytorch'
            )
            
            # Save model
            model_path = os.path.join(OUTDIR, 'pytorch_models', f'pytorch_model_{run_counter}.pth')
            torch.save({
                'model_state_dict': model.state_dict(),
                'params': params,
                'test_acc': test_acc,
                'test_metrics': test_metrics,
                'cv_val_mean': cv_val_mean if use_cross_validation else None,
                'run': run_counter
            }, model_path)
            
            # Track best model
            score_for_comparison = cv_val_mean if use_cross_validation else test_acc
            if score_for_comparison and score_for_comparison > best_test_acc:
                best_test_acc = score_for_comparison
                best_pytorch_model = {
                    'model': model,
                    'experiment_id': run_counter,
                    'params': params,
                    'test_metrics': test_metrics,
                    'train_metrics': train_metrics,
                    'model_path': model_path
                }
            
            # Log cross-validation metrics to MLflow
            if use_mlflow and use_cross_validation:
                if cv_train_mean is not None:
                    mlflow.log_metric("cv_train_mean", float(cv_train_mean))
                if cv_train_std is not None:
                    mlflow.log_metric("cv_train_std", float(cv_train_std))
                if cv_val_mean is not None:
                    mlflow.log_metric("cv_val_mean", float(cv_val_mean))
                if cv_val_std is not None:
                    mlflow.log_metric("cv_val_std", float(cv_val_std))
                mlflow.log_metric("final_train_acc", float(train_acc))
                mlflow.log_metric("final_test_acc", float(test_acc))
            
            # Store results
            result = {
                'run': run_counter,
                'params': params,
                'train_acc': train_acc,
                'test_acc': test_acc,
                'train_precision_macro': train_metrics['precision_macro'],
                'train_recall_macro': train_metrics['recall_macro'],
                'train_f1_macro': train_metrics['f1_macro'],
                'test_precision_macro': test_metrics['precision_macro'],
                'test_recall_macro': test_metrics['recall_macro'],
                'test_f1_macro': test_metrics['f1_macro'],
                'test_precision_weighted': test_metrics['precision_weighted'],
                'test_recall_weighted': test_metrics['recall_weighted'],
                'test_f1_weighted': test_metrics['f1_weighted'],
                'cv_train_mean': cv_train_mean,
                'cv_train_std': cv_train_std,
                'cv_val_mean': cv_val_mean,
                'cv_val_std': cv_val_std,
                'cv_loss_mean': cv_loss_mean,
                'final_train_loss': train_losses[-1] if train_losses else None,
                'final_val_loss': val_losses[-1] if val_losses else None,
                'best_val_loss': early_stopping.val_loss_min,
                'early_stopped': early_stopping.early_stop,
                'epochs_trained': len(train_losses),  # Actual epochs trained (may be less due to early stopping)
                'early_stop_epoch': len(train_losses) if early_stopping.early_stop else None,
                'duration_seconds': duration,
                'timestamp': start_time.isoformat(),
                'model_path': model_path
            }
            
            pytorch_results.append(result)
            
            # Confusion matrix logging
            cm = confusion_matrix(y_test, pred_test)
            fig, ax = plt.subplots(figsize=(8, 6))
            sns.heatmap(cm, annot=True, fmt='d', ax=ax, cmap='Blues')
            ax.set_title(f'Confusion Matrix - Run {run_counter}')
            ax.set_xlabel('Predicted')
            ax.set_ylabel('Actual')
            writer.add_figure('Confusion_Matrix', fig, global_step=run_counter)
            plt.close(fig)
            
            # Training loss curve plot with fixed dimension matching
            fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 5))
            
            # Plot training losses (one per epoch)
            ax1.plot(range(len(train_losses)), train_losses, label='Training Loss')
            
            # Plot validation losses with correct epochs alignment
            if val_losses and len(val_losses) > 0:
                # val_epochs contains the actual epochs where validation occurred
                # val_losses contains the corresponding validation losses
                ax1.plot(val_epochs, val_losses, label='Validation Loss', marker='o')
            
            ax1.set_title(f"Loss Curves - Run {run_counter}")
            ax1.set_xlabel('Epoch')
            ax1.set_ylabel('Loss')
            ax1.legend()
            ax1.grid(True)
            
            # Plot validation accuracies
            if val_accuracies and len(val_accuracies) > 0:
                ax2.plot(val_epochs, val_accuracies, label='Validation Accuracy', marker='o', color='green')
                ax2.set_title(f"Validation Accuracy - Run {run_counter}")
                ax2.set_xlabel('Epoch')
                ax2.set_ylabel('Accuracy')
                ax2.legend()
                ax2.grid(True)
            
            plt.tight_layout()
            
            # Save plot
            fname = f"pytorch_metrics_run_{run_counter}.png"
            plt.savefig(os.path.join(OUTDIR, fname), dpi=300, bbox_inches='tight')
            plt.close()
            
            # Progress logging
            if use_cross_validation:
                logger.info(f"PyTorch experiment {run_counter}/{len(combos)} completed. CV val: {cv_val_mean:.4f}±{cv_val_std:.4f}, Test F1: {test_metrics['f1_macro']:.4f}")
            else:
                logger.info(f"PyTorch experiment {run_counter}/{len(combos)} completed. Train acc: {train_acc:.4f}, Test F1: {test_metrics['f1_macro']:.4f}")
            
            # Clean up
            writer.close()
            if use_mlflow:
                mlflow.end_run()
            
            # Memory cleanup
            del model
            if torch.cuda.is_available():
                torch.cuda.empty_cache()
        
        except Exception as e:
            logger.error(f"PyTorch experiment {run_counter} failed: {str(e)}")
            logger.error(traceback.format_exc())
            if use_mlflow:
                mlflow.end_run(status="FAILED")
            continue
    
    logger.info(f"Completed {len(pytorch_results)} successful PyTorch experiments")
    
    # Save best PyTorch model with special designation
    if best_pytorch_model:
        best_model_dir = os.path.join(OUTDIR, 'best_models')
        os.makedirs(best_model_dir, exist_ok=True)
        
        # Save best model
        best_model_path = os.path.join(best_model_dir, 'best_pytorch_model.pth')
        torch.save({
            'model_state_dict': best_pytorch_model['model'].state_dict(),
            'params': best_pytorch_model['params'],
            'test_metrics': best_pytorch_model['test_metrics'],
            'run': best_pytorch_model['experiment_id']
        }, best_model_path)
        
        # Save best model metrics with special naming
        with torch.no_grad():
            best_pred = best_pytorch_model['model'](X_test_t.to(device)).argmax(dim=1).cpu().numpy()
        
        best_test_metrics = calculate_comprehensive_metrics(
            y_test, best_pred, 'BEST_pytorch_model', 'BEST', 'pytorch_best'
        )
        
        # Save best model info
        best_info = {
            'experiment_id': best_pytorch_model['experiment_id'],
            'params': best_pytorch_model['params'],
            'metrics': best_test_metrics,
            'model_path': best_model_path,
            'timestamp': datetime.now().isoformat()
        }
        
        with open(os.path.join(best_model_dir, 'best_pytorch_model_info.json'), 'w') as f:
            json.dump(best_info, f, indent=2)
        
        logger.info(f"Best PyTorch model saved (Experiment {best_pytorch_model['experiment_id']}) with score: {best_test_acc:.4f}")
    
    # Save PyTorch results
    if pytorch_results:
        pytorch_df = pd.DataFrame([
            {
                'run': r['run'],
                'hidden_sizes': str(r['params']['hidden_sizes']),
                'learning_rate': r['params']['learning_rate'],
                'dropout': r['params']['dropout'],
                'batch_size': r['params']['batch_size'],
                'epochs': r['params']['epochs'],
                'train_acc': r['train_acc'],
                'test_acc': r['test_acc'],
                'train_f1_macro': r['train_f1_macro'],
                'test_f1_macro': r['test_f1_macro'],
                'test_precision_macro': r['test_precision_macro'],
                'test_recall_macro': r['test_recall_macro'],
                'test_f1_weighted': r['test_f1_weighted'],
                'cv_train_mean': r['cv_train_mean'],
                'cv_train_std': r['cv_train_std'],
                'cv_val_mean': r['cv_val_mean'],
                'cv_val_std': r['cv_val_std'],
                'cv_loss_mean': r['cv_loss_mean'],
                'final_train_loss': r['final_train_loss'],
                'final_val_loss': r['final_val_loss'],
                'best_val_loss': r.get('best_val_loss', None),
                'early_stopped': r.get('early_stopped', False),
                'epochs_trained': r.get('epochs_trained', r['params']['epochs']),
                'early_stop_epoch': r.get('early_stop_epoch', None),
                'duration_seconds': r['duration_seconds'],
                'timestamp': r['timestamp'],
                'model_path': r['model_path']
            } for r in pytorch_results
        ])
        pytorch_df.to_csv(os.path.join(OUTDIR, 'pytorch_results_summary.csv'), index=False)
        logger.info("PyTorch results saved to pytorch_results_summary.csv")
        
        # Display best results
        if use_cross_validation:
            best_pytorch = pytorch_df.loc[pytorch_df['cv_val_mean'].idxmax()]
            print(f"\nBest PyTorch result (by CV score):")
            print(f"CV val accuracy: {best_pytorch['cv_val_mean']:.4f}±{best_pytorch['cv_val_std']:.4f}")
            print(f"Test accuracy: {best_pytorch['test_acc']:.4f}")
            print(f"Test F1-macro: {best_pytorch['test_f1_macro']:.4f}")
            print(f"Test Precision-macro: {best_pytorch['test_precision_macro']:.4f}")
            print(f"Test Recall-macro: {best_pytorch['test_recall_macro']:.4f}")
            print(f"Parameters: hidden_sizes={best_pytorch['hidden_sizes']}, lr={best_pytorch['learning_rate']}, dropout={best_pytorch['dropout']}")
        else:
            best_pytorch = pytorch_df.loc[pytorch_df['test_acc'].idxmax()]
            print(f"\nBest PyTorch result:")
            print(f"Test accuracy: {best_pytorch['test_acc']:.4f}")
            print(f"Test F1-macro: {best_pytorch['test_f1_macro']:.4f}")
            print(f"Test Precision-macro: {best_pytorch['test_precision_macro']:.4f}")
            print(f"Test Recall-macro: {best_pytorch['test_recall_macro']:.4f}")
            print(f"Parameters: hidden_sizes={best_pytorch['hidden_sizes']}, lr={best_pytorch['learning_rate']}, dropout={best_pytorch['dropout']}")

else:
    logger.info("PyTorch experiments skipped")

2025-09-06 00:02:05,157 - INFO - Starting 4 PyTorch experiments
2025-09-06 00:02:05,158 - INFO - Using 5-fold cross-validation (this will increase training time by ~5x)
PyTorch experiments:   0%|          | 0/4 [00:00<?, ?it/s]2025-09-06 00:02:05,161 - INFO - 
PyTorch experiment 1/4: {'hidden_sizes': [64], 'learning_rate': 0.001, 'dropout': 0.2, 'batch_size': 64, 'epochs': 100}
2025-09-06 00:02:05,161 - INFO - 
PyTorch experiment 1/4: {'hidden_sizes': [64], 'learning_rate': 0.001, 'dropout': 0.2, 'batch_size': 64, 'epochs': 100}
2025-09-06 00:04:11,243 - INFO - Early stopping triggered at epoch 52/100 (patience=10)
2025-09-06 00:04:11,244 - INFO - Best validation loss: 0.020536
2025-09-06 00:04:11,245 - INFO - Restored model to best validation loss: 0.020536
2025-09-06 00:04:11,245 - INFO - Model weights restored to best validation loss checkpoint
2025-09-06 00:04:11,243 - INFO - Early stopping triggered at epoch 52/100 (patience=10)
2025-09-06 00:04:11,244 - INFO - Best validation los


Best PyTorch result (by CV score):
CV val accuracy: 0.9930±0.0016
Test accuracy: 0.9930
Test F1-macro: 0.9930
Test Precision-macro: 0.9934
Test Recall-macro: 0.9927
Parameters: hidden_sizes=[64], lr=0.001, dropout=0.2


In [11]:
# Comprehensive Results Analysis with All Metrics
logger.info("Starting comprehensive analysis and plotting of all metrics")

def create_comprehensive_performance_plots(df, model_type):
    """Create comprehensive performance plots for all metrics"""
    
    # Prepare data for plotting
    if 'hidden_layer_sizes' in df.columns:
        df['arch_str'] = df['hidden_layer_sizes'].apply(lambda x: str(x).replace(' ', ''))
        x_col = 'arch_str'
        hue_col = 'solver'
    else:
        df['arch_str'] = df['hidden_sizes'].apply(lambda x: str(x).replace(' ', ''))
        x_col = 'arch_str'
        hue_col = 'dropout'
    
    # Create a 2x2 subplot for different metrics
    fig, axes = plt.subplots(2, 2, figsize=(20, 15))
    fig.suptitle(f'{model_type} - Comprehensive Performance Analysis', fontsize=16)
    
    # Plot 1: Accuracy
    sns.barplot(x=x_col, y='test_acc', hue=hue_col, data=df, ax=axes[0,0], palette='viridis')
    axes[0,0].set_title('Test Accuracy')
    axes[0,0].set_xlabel('Architecture')
    axes[0,0].set_ylabel('Accuracy')
    axes[0,0].tick_params(axis='x', rotation=45)
    axes[0,0].grid(axis='y', linestyle='--', alpha=0.7)
    
    # Plot 2: F1-Score (Macro)
    sns.barplot(x=x_col, y='test_f1_macro', hue=hue_col, data=df, ax=axes[0,1], palette='plasma')
    axes[0,1].set_title('Test F1-Score (Macro)')
    axes[0,1].set_xlabel('Architecture')
    axes[0,1].set_ylabel('F1-Score')
    axes[0,1].tick_params(axis='x', rotation=45)
    axes[0,1].grid(axis='y', linestyle='--', alpha=0.7)
    
    # Plot 3: Precision (Macro)
    sns.barplot(x=x_col, y='test_precision_macro', hue=hue_col, data=df, ax=axes[1,0], palette='cividis')
    axes[1,0].set_title('Test Precision (Macro)')
    axes[1,0].set_xlabel('Architecture')
    axes[1,0].set_ylabel('Precision')
    axes[1,0].tick_params(axis='x', rotation=45)
    axes[1,0].grid(axis='y', linestyle='--', alpha=0.7)
    
    # Plot 4: Recall (Macro)
    sns.barplot(x=x_col, y='test_recall_macro', hue=hue_col, data=df, ax=axes[1,1], palette='magma')
    axes[1,1].set_title('Test Recall (Macro)')
    axes[1,1].set_xlabel('Architecture')
    axes[1,1].set_ylabel('Recall')
    axes[1,1].tick_params(axis='x', rotation=45)
    axes[1,1].grid(axis='y', linestyle='--', alpha=0.7)
    
    plt.tight_layout()
    filename = f'{model_type.lower()}_comprehensive_metrics.png'
    plt.savefig(os.path.join(OUTDIR, filename), dpi=300, bbox_inches='tight')
    plt.close()
    logger.info(f"Saved comprehensive metrics plot: {filename}")

def create_cv_comparison_plots(df, model_type):
    """Create cross-validation comparison plots if CV was used"""
    if not use_cross_validation:
        return
    
    cv_col = 'cv_mean_acc' if model_type == 'Scikit-learn' else 'cv_val_mean'
    if cv_col not in df.columns:
        return
    
    # Prepare data for plotting
    if 'hidden_layer_sizes' in df.columns:
        df['arch_str'] = df['hidden_layer_sizes'].apply(lambda x: str(x).replace(' ', ''))
        x_col = 'arch_str'
        hue_col = 'solver'
    else:
        df['arch_str'] = df['hidden_sizes'].apply(lambda x: str(x).replace(' ', ''))
        x_col = 'arch_str'
        hue_col = 'dropout'
    
    fig, axes = plt.subplots(1, 2, figsize=(15, 6))
    
    # CV results with error bars
    cv_std_col = cv_col.replace('mean', 'std')
    if cv_std_col in df.columns:
        # Group data for error bars
        grouped = df.groupby([x_col, hue_col]).agg({
            cv_col: 'mean',
            cv_std_col: 'mean'
        }).reset_index()
        
        sns.barplot(x=x_col, y=cv_col, hue=hue_col, data=grouped, ax=axes[0], palette='viridis')
        
        # Add error bars
        for i, (_, row) in enumerate(grouped.iterrows()):
            axes[0].errorbar(i, row[cv_col], yerr=row[cv_std_col], 
                           fmt='none', color='black', capsize=3, alpha=0.7)
        
        axes[0].set_title(f'{model_type} - Cross-Validation Results')
        axes[0].set_xlabel('Architecture')
        axes[0].set_ylabel('CV Accuracy')
        axes[0].tick_params(axis='x', rotation=45)
        axes[0].grid(axis='y', linestyle='--', alpha=0.7)
    
    # CV vs Test correlation
    axes[1].scatter(df[cv_col], df['test_acc'], alpha=0.7)
    axes[1].plot([0, 1], [0, 1], 'r--', alpha=0.8)
    correlation = df[cv_col].corr(df['test_acc'])
    axes[1].set_xlabel('Cross-Validation Accuracy')
    axes[1].set_ylabel('Test Set Accuracy')
    axes[1].set_title(f'{model_type} CV vs Test Correlation (r={correlation:.3f})')
    axes[1].grid(True, alpha=0.3)
    
    plt.tight_layout()
    filename = f'{model_type.lower()}_cv_analysis.png'
    plt.savefig(os.path.join(OUTDIR, filename), dpi=300, bbox_inches='tight')
    plt.close()
    logger.info(f"Saved CV analysis plot: {filename}")

# Create comprehensive plots for both model types
if 'sklearn_df' in locals() and not sklearn_df.empty:
    create_comprehensive_performance_plots(sklearn_df, 'Scikit-learn')
    create_cv_comparison_plots(sklearn_df, 'Scikit-learn')
    
    print("\n" + "="*60)
    print("SCIKIT-LEARN COMPREHENSIVE RESULTS SUMMARY")
    print("="*60)
    print(f"Total experiments: {len(sklearn_df)}")
    
    # Best results summary
    metrics_to_show = ['test_acc', 'test_f1_macro', 'test_precision_macro', 'test_recall_macro']
    if use_cross_validation:
        cv_best = sklearn_df.loc[sklearn_df['cv_mean_acc'].idxmax()]
        print(f"\nBest model (by CV accuracy):")
        print(f"  CV accuracy: {cv_best['cv_mean_acc']:.4f}±{cv_best['cv_std_acc']:.4f}")
        for metric in metrics_to_show:
            print(f"  {metric}: {cv_best[metric]:.4f}")
        print(f"  Parameters: {cv_best['hidden_layer_sizes']}, {cv_best['solver']}, lr={cv_best['learning_rate_init']}")
    
    # Overall statistics
    print(f"\nOverall Performance Statistics:")
    for metric in metrics_to_show:
        print(f"  {metric}: {sklearn_df[metric].mean():.4f}±{sklearn_df[metric].std():.4f} (best: {sklearn_df[metric].max():.4f})")

if 'pytorch_df' in locals() and not pytorch_df.empty:
    create_comprehensive_performance_plots(pytorch_df, 'PyTorch')
    create_cv_comparison_plots(pytorch_df, 'PyTorch')
    
    print("\n" + "="*60)
    print("PYTORCH COMPREHENSIVE RESULTS SUMMARY")
    print("="*60)
    print(f"Total experiments: {len(pytorch_df)}")
    
    # Best results summary
    metrics_to_show = ['test_acc', 'test_f1_macro', 'test_precision_macro', 'test_recall_macro']
    if use_cross_validation:
        cv_best = pytorch_df.loc[pytorch_df['cv_val_mean'].idxmax()]
        print(f"\nBest model (by CV accuracy):")
        print(f"  CV accuracy: {cv_best['cv_val_mean']:.4f}±{cv_best['cv_val_std']:.4f}")
        for metric in metrics_to_show:
            print(f"  {metric}: {cv_best[metric]:.4f}")
        print(f"  Parameters: {cv_best['hidden_sizes']}, lr={cv_best['learning_rate']}, dropout={cv_best['dropout']}")
    
    # Overall statistics  
    print(f"\nOverall Performance Statistics:")
    for metric in metrics_to_show:
        print(f"  {metric}: {pytorch_df[metric].mean():.4f}±{pytorch_df[metric].std():.4f} (best: {pytorch_df[metric].max():.4f})")

# Model Comparison (if both models were run)
if 'sklearn_df' in locals() and 'pytorch_df' in locals() and not sklearn_df.empty and not pytorch_df.empty:
    print("\n" + "="*60)
    print("MODEL COMPARISON SUMMARY")
    print("="*60)
    
    comparison_metrics = ['test_acc', 'test_f1_macro', 'test_precision_macro', 'test_recall_macro']
    
    print("Best Performance Comparison:")
    for metric in comparison_metrics:
        sklearn_best = sklearn_df[metric].max()
        pytorch_best = pytorch_df[metric].max()
        winner = "Scikit-learn" if sklearn_best > pytorch_best else "PyTorch"
        print(f"  {metric}: Sklearn={sklearn_best:.4f}, PyTorch={pytorch_best:.4f} → Winner: {winner}")
    
    print("\nAverage Performance Comparison:")
    for metric in comparison_metrics:
        sklearn_avg = sklearn_df[metric].mean()
        pytorch_avg = pytorch_df[metric].mean()
        winner = "Scikit-learn" if sklearn_avg > pytorch_avg else "PyTorch"
        print(f"  {metric}: Sklearn={sklearn_avg:.4f}, PyTorch={pytorch_avg:.4f} → Winner: {winner}")

# Training time analysis
print("\n" + "="*60)
print("TRAINING TIME AND COMPUTATIONAL ANALYSIS")
print("="*60)

if 'sklearn_df' in locals() and not sklearn_df.empty:
    sklearn_time = sklearn_df['duration_seconds']
    print(f"Scikit-learn:")
    print(f"  Average time per experiment: {sklearn_time.mean():.2f}s")
    print(f"  Total time: {sklearn_time.sum():.2f}s ({sklearn_time.sum()/60:.1f} minutes)")

if 'pytorch_df' in locals() and not pytorch_df.empty:
    pytorch_time = pytorch_df['duration_seconds']
    print(f"PyTorch:")
    print(f"  Average time per experiment: {pytorch_time.mean():.2f}s")
    print(f"  Total time: {pytorch_time.sum():.2f}s ({pytorch_time.sum()/60:.1f} minutes)")

if use_cross_validation:
    print(f"\nCross-validation impact:")
    print(f"  Estimated {n_folds}x time increase due to CV")
    if 'sklearn_df' in locals():
        print(f"  Sklearn estimated time without CV: {sklearn_df['duration_seconds'].mean() / n_folds:.2f}s per experiment")
    if 'pytorch_df' in locals():
        print(f"  PyTorch estimated time without CV: {pytorch_df['duration_seconds'].mean() / n_folds:.2f}s per experiment")

print("\n" + "="*60)
print("FILES AND OUTPUTS SUMMARY")
print("="*60)
print("Generated files and directories:")
print("📁 experiment_outputs/")
print("  ├── 📁 sklearn_models/ - Individual sklearn models (.joblib)")
print("  ├── 📁 pytorch_models/ - Individual PyTorch models (.pth)")
print("  ├── 📁 best_models/ - Best models from each framework")
print("  ├── 📁 metrics/ - Comprehensive metrics for each model (.json)")
print("  ├── 📁 confusion_matrices/ - Confusion matrix plots (.png)")
print("  ├── 📁 tb_logs/ - TensorBoard logs (PyTorch only)")
print("  ├── 📄 sklearn_results_summary.csv - Sklearn experiment results")
print("  ├── 📄 pytorch_results_summary.csv - PyTorch experiment results")
print("  ├── 📄 experiment_config.yaml - Experiment configuration")
print("  └── 📄 *.png - Various performance plots")

logger.info("Comprehensive analysis complete.")

2025-09-06 00:10:11,098 - INFO - Starting comprehensive analysis and plotting of all metrics
2025-09-06 00:10:12,071 - INFO - Saved comprehensive metrics plot: pytorch_comprehensive_metrics.png
2025-09-06 00:10:12,071 - INFO - Saved comprehensive metrics plot: pytorch_comprehensive_metrics.png
2025-09-06 00:10:12,534 - INFO - Saved CV analysis plot: pytorch_cv_analysis.png
2025-09-06 00:10:12,537 - INFO - Comprehensive analysis complete.
2025-09-06 00:10:12,534 - INFO - Saved CV analysis plot: pytorch_cv_analysis.png
2025-09-06 00:10:12,537 - INFO - Comprehensive analysis complete.



PYTORCH COMPREHENSIVE RESULTS SUMMARY
Total experiments: 4

Best model (by CV accuracy):
  CV accuracy: 0.9930±0.0016
  test_acc: 0.9930
  test_f1_macro: 0.9930
  test_precision_macro: 0.9934
  test_recall_macro: 0.9927
  Parameters: [64], lr=0.001, dropout=0.2

Overall Performance Statistics:
  test_acc: 0.9927±0.0009 (best: 0.9934)
  test_f1_macro: 0.9928±0.0009 (best: 0.9936)
  test_precision_macro: 0.9933±0.0009 (best: 0.9942)
  test_recall_macro: 0.9923±0.0010 (best: 0.9931)

TRAINING TIME AND COMPUTATIONAL ANALYSIS
PyTorch:
  Average time per experiment: 31.37s
  Total time: 125.49s (2.1 minutes)

Cross-validation impact:
  Estimated 5x time increase due to CV
  PyTorch estimated time without CV: 6.27s per experiment

FILES AND OUTPUTS SUMMARY
Generated files and directories:
📁 experiment_outputs/
  ├── 📁 sklearn_models/ - Individual sklearn models (.joblib)
  ├── 📁 pytorch_models/ - Individual PyTorch models (.pth)
  ├── 📁 best_models/ - Best models from each framework
  ├── 📁 

In [16]:
# ======================================
# DEPLOYMENT PACKAGE CREATION (FIXED)
# ======================================

import os
import json
import joblib
from sklearn.preprocessing import StandardScaler

# Create deployment directory
best_model_dir = os.path.join(OUTDIR, 'best_models')
os.makedirs(best_model_dir, exist_ok=True)

print("🚀 CREATING FIXED DEPLOYMENT PACKAGE...")
print("="*60)

# Extract feature names for app deployment
feature_names = list(X.columns)  # The selected features used in training (in correct order)
n_features = len(X.columns)

print(f"📊 FEATURE INFORMATION:")
print(f"   Selected features for training: {n_features}")
print(f"   Feature names: {feature_names}")

# Save feature names as pickle/joblib file (for app requirements)
feature_names_path = os.path.join(best_model_dir, 'feature_names.pkl')
joblib.dump(feature_names, feature_names_path)

print(f"✅ Feature names saved to: {feature_names_path}")

# ⚠️ CRITICAL FIX: Create scaler fitted ONLY on selected features
print(f"\n🔧 CREATING CORRECT SCALER...")
print(f"   Original data shape: {df.shape}")
print(f"   Selected features shape: {X.shape}")

# Create and fit scaler on the SAME selected features used for training
correct_scaler = StandardScaler()
X_selected_features = df[feature_names]  # Select features from original data
correct_scaler.fit(X_selected_features)

print(f"   Scaler fitted on: {correct_scaler.n_features_in_} features ✅")
print(f"   Expected input shape: {correct_scaler.n_features_in_}")

# Save the CORRECT scaler
scaler_path = os.path.join(best_model_dir, 'scaler.pkl')
joblib.dump(correct_scaler, scaler_path)

print(f"✅ CORRECT scaler saved to: {scaler_path}")

# Also save as joblib format
scaler_joblib_path = os.path.join(best_model_dir, 'scaler.joblib')
joblib.dump(correct_scaler, scaler_joblib_path)

print(f"✅ CORRECT scaler (joblib) saved to: {scaler_joblib_path}")

# Feature names and preprocessing information (comprehensive metadata)
feature_info = {
    'feature_names': feature_names,
    'n_features': n_features,
    'feature_selection_info': {
        'feature_selection_applied': True,
        'selected_features': selected_features,  # Original feature selection list
        'available_features': available_features,  # Features that were actually available
        'missing_features': missing_features if 'missing_features' in locals() else [],
        'total_original_features': len(selected_features)
    },
    'preprocessing_info': {
        'scaler_type': 'StandardScaler',
        'scaling_applied': True,
        'scaler_fitted_on_features': n_features,  # CRITICAL: scaler expects exactly these features
        'target_column': target_col,
        'data_split_info': {
            'train_size': X_train.shape[0],
            'test_size': X_test.shape[0],
            'train_ratio': X_train.shape[0] / (X_train.shape[0] + X_test.shape[0])
        }
    }
}

# Save feature information
feature_info_path = os.path.join(best_model_dir, 'feature_info.json')
with open(feature_info_path, 'w') as f:
    json.dump(feature_info, f, indent=2)

print(f"✅ Feature information saved to: {feature_info_path}")

# Update best model info files with correct scaler paths
if 'best_pytorch_model' in locals() and best_pytorch_model:
    # Update PyTorch model info
    pytorch_info_path = os.path.join(best_model_dir, 'best_pytorch_model_info.json')
    if os.path.exists(pytorch_info_path):
        with open(pytorch_info_path, 'r') as f:
            pytorch_best_info = json.load(f)
        
        # Add scaler and feature info
        pytorch_best_info['deployment_files'] = {
            'feature_names_pkl': feature_names_path,
            'scaler_pkl': scaler_path,
            'scaler_joblib': scaler_joblib_path,
            'feature_info_json': feature_info_path
        }
        
        with open(pytorch_info_path, 'w') as f:
            json.dump(pytorch_best_info, f, indent=2)
        
        print(f"✅ Updated PyTorch best model info with correct scaler paths")

# Create deployment summary
deployment_summary = {
    'deployment_ready': True,
    'model_type': 'PyTorch' if 'best_pytorch_model' in locals() and best_pytorch_model else 'Scikit-learn',
    'feature_count': n_features,
    'scaler_input_features': correct_scaler.n_features_in_,
    'files': {
        'feature_names': 'feature_names.pkl',
        'scaler': 'scaler.pkl',
        'scaler_joblib': 'scaler.joblib', 
        'feature_info': 'feature_info.json',
        'model': 'best_pytorch_model.pth' if 'best_pytorch_model' in locals() and best_pytorch_model else 'best_sklearn_model.joblib'
    },
    'usage_instructions': {
        'step_1': 'Load feature_names.pkl to get the exact feature list',
        'step_2': 'Ensure your input data has exactly these features in this order',
        'step_3': 'Load scaler.pkl and apply: X_scaled = scaler.transform(X)',
        'step_4': 'Load model and predict: predictions = model.predict(X_scaled)',
        'critical_note': f'Input data MUST have exactly {n_features} features matching the feature_names list'
    }
}

deployment_summary_path = os.path.join(best_model_dir, 'deployment_summary.json')
with open(deployment_summary_path, 'w') as f:
    json.dump(deployment_summary, f, indent=2)

print(f"✅ Deployment summary saved to: {deployment_summary_path}")

# Test the deployment package
print(f"\n🧪 TESTING DEPLOYMENT PACKAGE...")
try:
    # Test loading all files
    test_feature_names = joblib.load(feature_names_path)
    test_scaler = joblib.load(scaler_path)
    
    print(f"✅ Test: Feature names loaded ({len(test_feature_names)} features)")
    print(f"✅ Test: Scaler loaded (expects {test_scaler.n_features_in_} features)")
    
    # Test with sample data
    sample_data = X.iloc[:1].copy()  # Take one row
    sample_scaled = test_scaler.transform(sample_data)
    
    print(f"✅ Test: Sample data transformation successful")
    print(f"   Input shape: {sample_data.shape}")
    print(f"   Output shape: {sample_scaled.shape}")
    
except Exception as e:
    print(f"❌ Test failed: {e}")

print(f"\n📦 DEPLOYMENT PACKAGE READY:")
print(f"Location: {best_model_dir}")
print(f"Files created:")
print(f"  ├── feature_names.pkl (List of {n_features} feature names - REQUIRED)")
print(f"  ├── scaler.pkl (StandardScaler fitted on {n_features} features)")  
print(f"  ├── scaler.joblib (Same scaler in joblib format)")
print(f"  ├── feature_info.json (Complete feature metadata)")
print(f"  ├── deployment_summary.json (Usage instructions)")
print(f"  └── Model files (best_pytorch_model.pth or best_sklearn_model.joblib)")

print(f"\n⚠️  CRITICAL FOR YOUR APP:")
print(f"   1. Your app data MUST have exactly {n_features} features")
print(f"   2. Feature names must match exactly: {feature_names}")
print(f"   3. Use the scaler.pkl file (fitted on {n_features} features)")
print(f"   4. Apply feature selection BEFORE scaling in your app")

logger.info("Fixed deployment package created successfully!")

2025-09-06 00:23:41,565 - INFO - Fixed deployment package created successfully!


🚀 CREATING FIXED DEPLOYMENT PACKAGE...
📊 FEATURE INFORMATION:
   Selected features for training: 35
   Feature names: ['has_dental_data', 'special_needs', 'caregiver_treatment', 'plaque', 'dry_mouth', 'enamel_defects', 'fluoride_water', 'fluoride_toothpaste', 'topical_fluoride', 'regular_checkups', 'sealed_pits', 'enamel_change', 'dentin_discoloration', 'white_spot_lesions', 'cavitated_lesions', 'multiple_restorations', 'missing_teeth', 'total_dmft_score', 'sweet_sugary_foods', 'sweet_sugary_foods_bedtime', 'takeaways_processed_foods', 'fresh_fruit', 'cold_drinks_juices', 'processed_fruit', 'processed_fruit_bedtime', 'spreads', 'spreads_bedtime', 'added_sugars', 'added_sugars_bedtime', 'salty_snacks', 'dairy_products', 'vegetables', 'has_dietary_data', 'cold_drinks_juices_bedtime', 'water']
✅ Feature names saved to: experiment_outputs\best_models\feature_names.pkl

🔧 CREATING CORRECT SCALER...
   Original data shape: (25000, 69)
   Selected features shape: (25000, 35)
   Scaler fitted 

In [17]:
# =======================================
# COMPREHENSIVE MODEL DEPLOYMENT PACKAGE
# =======================================

import os
import json
import joblib
import torch
import pickle
from sklearn.preprocessing import StandardScaler

print("🚀 CREATING COMPREHENSIVE DEPLOYMENT PACKAGE")
print("="*70)

# Load the best PyTorch model
pytorch_model_path = os.path.join(best_model_dir, 'best_pytorch_model.pth')
pytorch_model_dict = torch.load(pytorch_model_path, map_location='cpu')

print("📦 SAVING MODEL IN MULTIPLE FORMATS:")

# 1. Save as Pickle (.pkl)
pytorch_pkl_path = os.path.join(best_model_dir, 'best_pytorch_model.pkl')
joblib.dump(pytorch_model_dict, pytorch_pkl_path)
print(f"✅ PyTorch model saved as pickle: best_pytorch_model.pkl")

# 2. Save using standard pickle
pytorch_pickle_path = os.path.join(best_model_dir, 'best_pytorch_model.pickle')
with open(pytorch_pickle_path, 'wb') as f:
    pickle.dump(pytorch_model_dict, f)
print(f"✅ PyTorch model saved as pickle: best_pytorch_model.pickle")

# 3. Save as Joblib
pytorch_joblib_path = os.path.join(best_model_dir, 'best_pytorch_model.joblib')
joblib.dump(pytorch_model_dict, pytorch_joblib_path)
print(f"✅ PyTorch model saved as joblib: best_pytorch_model.joblib")

# 4. Keep original .pth format
print(f"✅ PyTorch model available as: best_pytorch_model.pth")

print(f"\n🔧 MODEL ARCHITECTURE & PARAMETERS:")
print(f"   Hidden layers: {pytorch_model_dict['params']['hidden_sizes']}")
print(f"   Input features: 35")
print(f"   Output classes: 3")
print(f"   Dropout: {pytorch_model_dict['params']['dropout']}")
print(f"   Learning rate: {pytorch_model_dict['params']['learning_rate']}")

print(f"\n📊 MODEL PERFORMANCE:")
metrics = pytorch_model_dict['test_metrics']
print(f"   Test Accuracy: {metrics['accuracy']:.4f}")
print(f"   F1-Score (macro): {metrics['f1_macro']:.4f}")
print(f"   Precision (macro): {metrics['precision_macro']:.4f}")
print(f"   Recall (macro): {metrics['recall_macro']:.4f}")

# Create a simple Python class for easy model loading and prediction
model_wrapper_code = '''
import torch
import torch.nn as nn
import joblib
import numpy as np

class OralHealthMLPWrapper:
    """
    Wrapper class for the OralHealth MLP model for easy deployment.
    """
    
    def __init__(self, model_path, scaler_path, feature_names_path):
        """
        Initialize the model wrapper.
        
        Args:
            model_path: Path to the saved model (.pkl, .pickle, .joblib, or .pth)
            scaler_path: Path to the saved scaler (.pkl or .joblib)
            feature_names_path: Path to the saved feature names (.pkl)
        """
        # Load feature names
        self.feature_names = joblib.load(feature_names_path)
        self.n_features = len(self.feature_names)
        
        # Load scaler
        self.scaler = joblib.load(scaler_path)
        
        # Load model
        if model_path.endswith('.pth'):
            self.model_dict = torch.load(model_path, map_location='cpu')
        else:
            self.model_dict = joblib.load(model_path)
        
        # Recreate the model architecture
        self.model = self._create_model()
        self.model.load_state_dict(self.model_dict['model_state_dict'])
        self.model.eval()
        
        # Class names
        self.class_names = ['Low Risk', 'Medium Risk', 'High Risk']
        
    def _create_model(self):
        """Recreate the model architecture."""
        hidden_sizes = self.model_dict['params']['hidden_sizes']
        dropout = self.model_dict['params']['dropout']
        
        layers = []
        input_size = self.n_features
        
        for hidden_size in hidden_sizes:
            layers.extend([
                nn.Linear(input_size, hidden_size),
                nn.ReLU(),
                nn.Dropout(dropout)
            ])
            input_size = hidden_size
        
        # Output layer
        layers.append(nn.Linear(input_size, 3))  # 3 classes
        
        return nn.Sequential(*layers)
    
    def predict(self, X):
        """
        Make predictions on input data.
        
        Args:
            X: Input data (pandas DataFrame or numpy array)
            
        Returns:
            predictions: Array of predicted class indices
        """
        # Ensure X has the correct features
        if hasattr(X, 'columns'):
            # DataFrame - select and reorder features
            X = X[self.feature_names]
        
        # Convert to numpy if needed
        if hasattr(X, 'values'):
            X = X.values
        
        # Scale the data
        X_scaled = self.scaler.transform(X)
        
        # Convert to tensor
        X_tensor = torch.FloatTensor(X_scaled)
        
        # Make predictions
        with torch.no_grad():
            outputs = self.model(X_tensor)
            predictions = torch.argmax(outputs, dim=1)
        
        return predictions.numpy()
    
    def predict_proba(self, X):
        """
        Get prediction probabilities.
        
        Args:
            X: Input data (pandas DataFrame or numpy array)
            
        Returns:
            probabilities: Array of prediction probabilities
        """
        # Ensure X has the correct features
        if hasattr(X, 'columns'):
            X = X[self.feature_names]
        
        if hasattr(X, 'values'):
            X = X.values
        
        # Scale the data
        X_scaled = self.scaler.transform(X)
        X_tensor = torch.FloatTensor(X_scaled)
        
        # Get probabilities
        with torch.no_grad():
            outputs = self.model(X_tensor)
            probabilities = torch.softmax(outputs, dim=1)
        
        return probabilities.numpy()
    
    def predict_with_labels(self, X):
        """
        Get predictions with class labels.
        
        Args:
            X: Input data
            
        Returns:
            tuple: (predictions, class_labels, probabilities)
        """
        pred_indices = self.predict(X)
        pred_labels = [self.class_names[i] for i in pred_indices]
        probabilities = self.predict_proba(X)
        
        return pred_indices, pred_labels, probabilities

# Usage example:
# model = OralHealthMLPWrapper(
#     model_path='best_pytorch_model.pkl',
#     scaler_path='scaler.pkl', 
#     feature_names_path='feature_names.pkl'
# )
# predictions = model.predict(your_data)
# pred_indices, pred_labels, probabilities = model.predict_with_labels(your_data)
'''

# Save the wrapper class code
wrapper_path = os.path.join(best_model_dir, 'model_wrapper.py')
with open(wrapper_path, 'w') as f:
    f.write(model_wrapper_code)

print(f"✅ Model wrapper class saved: model_wrapper.py")

# Create deployment instructions
deployment_instructions = {
    "deployment_ready": True,
    "model_formats": {
        "pytorch_pth": "best_pytorch_model.pth (PyTorch format)",
        "pickle_pkl": "best_pytorch_model.pkl (Joblib pickle)",
        "pickle_std": "best_pytorch_model.pickle (Standard pickle)", 
        "joblib": "best_pytorch_model.joblib (Joblib format)"
    },
    "preprocessing": {
        "feature_names": "feature_names.pkl (35 features)",
        "scaler": "scaler.pkl (StandardScaler fitted on 35 features)"
    },
    "model_info": {
        "input_features": 35,
        "output_classes": 3,
        "class_names": ["Low Risk", "Medium Risk", "High Risk"],
        "architecture": pytorch_model_dict['params']
    },
    "usage_options": {
        "option_1_simple": {
            "description": "Load individual files manually",
            "steps": [
                "feature_names = joblib.load('feature_names.pkl')",
                "scaler = joblib.load('scaler.pkl')",
                "model_dict = joblib.load('best_pytorch_model.pkl')",
                "# Apply preprocessing and prediction manually"
            ]
        },
        "option_2_wrapper": {
            "description": "Use the provided wrapper class (RECOMMENDED)",
            "steps": [
                "from model_wrapper import OralHealthMLPWrapper",
                "model = OralHealthMLPWrapper('best_pytorch_model.pkl', 'scaler.pkl', 'feature_names.pkl')",
                "predictions = model.predict(your_data)",
                "pred_indices, pred_labels, probabilities = model.predict_with_labels(your_data)"
            ]
        }
    }
}

instructions_path = os.path.join(best_model_dir, 'deployment_instructions.json')
with open(instructions_path, 'w') as f:
    json.dump(deployment_instructions, f, indent=2)

print(f"✅ Deployment instructions saved: deployment_instructions.json")

print(f"\n📁 FINAL DEPLOYMENT PACKAGE:")
print(f"├── 🤖 MODEL FILES (Choose one):")
print(f"│   ├── best_pytorch_model.pkl     (Joblib pickle - RECOMMENDED)")
print(f"│   ├── best_pytorch_model.pickle  (Standard pickle)")
print(f"│   ├── best_pytorch_model.joblib  (Joblib format)")
print(f"│   └── best_pytorch_model.pth     (PyTorch format)")
print(f"├── 🔧 PREPROCESSING:")
print(f"│   ├── feature_names.pkl          (35 feature names)")
print(f"│   ├── scaler.pkl                 (StandardScaler)")
print(f"│   └── feature_info.json          (Feature metadata)")
print(f"├── 📝 DEPLOYMENT TOOLS:")
print(f"│   ├── model_wrapper.py           (Easy-to-use wrapper class)")
print(f"│   └── deployment_instructions.json (Complete usage guide)")
print(f"└── 📊 DOCUMENTATION:")
print(f"    └── deployment_summary.json    (Summary and metadata)")

print(f"\n🎯 RECOMMENDED USAGE:")
print(f"   1. Use: best_pytorch_model.pkl (easiest to load)")
print(f"   2. Use: model_wrapper.py (handles everything automatically)")
print(f"   3. Example: model = OralHealthMLPWrapper('best_pytorch_model.pkl', 'scaler.pkl', 'feature_names.pkl')")

# Test the pickle model loading
print(f"\n🧪 TESTING PICKLE MODEL LOADING:")
try:
    test_model = joblib.load(pytorch_pkl_path)
    print(f"✅ Pickle model loads successfully")
    print(f"   Model type: {type(test_model)}")
    print(f"   Contains keys: {list(test_model.keys())}")
except Exception as e:
    print(f"❌ Error loading pickle model: {e}")

logger.info("Comprehensive deployment package created with multiple model formats!")

2025-09-06 01:10:33,257 - INFO - Comprehensive deployment package created with multiple model formats!


🚀 CREATING COMPREHENSIVE DEPLOYMENT PACKAGE
📦 SAVING MODEL IN MULTIPLE FORMATS:
✅ PyTorch model saved as pickle: best_pytorch_model.pkl
✅ PyTorch model saved as pickle: best_pytorch_model.pickle
✅ PyTorch model saved as joblib: best_pytorch_model.joblib
✅ PyTorch model available as: best_pytorch_model.pth

🔧 MODEL ARCHITECTURE & PARAMETERS:
   Hidden layers: [64]
   Input features: 35
   Output classes: 3
   Dropout: 0.2
   Learning rate: 0.001

📊 MODEL PERFORMANCE:
   Test Accuracy: 0.9930
   F1-Score (macro): 0.9930
   Precision (macro): 0.9934
   Recall (macro): 0.9927
✅ Model wrapper class saved: model_wrapper.py
✅ Deployment instructions saved: deployment_instructions.json

📁 FINAL DEPLOYMENT PACKAGE:
├── 🤖 MODEL FILES (Choose one):
│   ├── best_pytorch_model.pkl     (Joblib pickle - RECOMMENDED)
│   ├── best_pytorch_model.pickle  (Standard pickle)
│   ├── best_pytorch_model.joblib  (Joblib format)
│   └── best_pytorch_model.pth     (PyTorch format)
├── 🔧 PREPROCESSING:
│   ├── fe